core/embed/lib/gl_color.h

```diff
@@ -0,0 +1,308 @@
+/*
+ * This file is part of the Trezor project, https://trezor.io/
+ *
+ * Copyright (c) SatoshiLabs
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef GL_COLOR_H
+#define GL_COLOR_H
+
+#include <stdint.h>
+
+#define GL_COLOR_16BIT
+// #define GL_COLOR_32BIT
+
+// Color in RGB565 format
+//
+// |15            8 | 7             0|
+// |---------------------------------|
+// |r r r r r g g g | g g g b b b b b|
+// |---------------------------------|
+
+typedef uint16_t gl_color16_t;
+
+// Color in RGBA8888 format
+//
+// |31           24 |23            16 |15             8 | 7             0 |
+// |----------------------------------------------------------------------|
+// |a a a a a a a a | r r r r r r r r | g g g g g g g g | b b b b b b b b |
+// |----------------------------------------------------------------------|
+//
+
+typedef uint32_t gl_color32_t;
+
+#ifdef GL_COLOR_16BIT
+#define gl_color_t gl_color16_t
+#define gl_color_to_color16(c) (c)
+#define gl_color16_to_color(c) (c)
+#define gl_color_to_color32(c) (gl_color16_to_color32(c))
+#define gl_color32_to_color(c) (gl_color32_to_color16(c))
+#define gl_color_lum(c) (gl_color16_lum(c))
+#elif GL_COLOR_32BIT
+#define gl_color_t gl_color32_t
+#define gl_color_to_color16(c) (gl_color32_to_color16(c))
+#define gl_color16_to_color(c) (gl_color16_to_color32(c))
+#define gl_color_to_color32(c) (c)
+#define gl_color32_to_color(c) (c)
+#else
+#error "GL_COLOR_16BIT/32BIT not specified"
+#endif
+
+// Constructs a 16-bit color from the given red (r),
+// green (g), and blue (b) values in the range 0..255
+static inline gl_color16_t gl_color16_rgb(uint8_t r, uint8_t g, uint8_t b) {
+  return ((r & 0xF8U) << 8) | ((g & 0xFCU) << 3) | ((b & 0xF8U) >> 3);
+}
+
+// Constructs a 32-bit color from the given red (r),
+// green (g), and blue (b) values in the range 0..255.
+// Alpha is set to 255.
+static inline gl_color32_t gl_color32_rgb(uint8_t r, uint8_t g, uint8_t b) {
+  return (0xFFU << 24) | ((uint32_t)r << 16) | ((uint32_t)g << 8) | b;
+}
+
+// Converts a 16-bit color to a 32-bit color; alpha is set to 255
+static inline gl_color32_t gl_color16_to_color32(gl_color16_t color) {
+  uint32_t r = (color & 0xF800) >> 8;
```

Comment:
I've refactored the code to avoid repeating the same calculation. Calculations are slightly different but with the same result. see 03b24c16d6c304d428ce5c5d2ba5463f5cc6ff94

Reply:
I think its nice improvement, could we do this also for the minority case when extracting the color in 5/6 bit format? just so that the code is more consistent.

There is no implementation of `gl_color32_blend_a8` for `GL_COLOR_32BIT`, is that intentional?

Reply:
I've added `gl_color32_blend_a8()` and `gl_color32_lum()`. For now, I'm leaving `gl_color16_blend_a4/a8()` as they are. I could use a macro to extract RGB components, but I would need to introduce a second one (just for these two places) to keep the efficiency at its current level.

see 0cee1236c82ac3db1fdca742c34e16846de07900